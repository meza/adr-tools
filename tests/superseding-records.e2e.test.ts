import * as childProcess from 'child_process';
import { realpathSync, rmdirSync } from 'node:fs';
import * as os from 'os';
import * as path from 'path';
import * as fs from 'fs/promises';
/* eslint-disable no-sync */
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

describe('Superseding Adrs', () => {
  const adr = path.resolve(path.dirname(__filename), '../src/index.ts');
  const command = `npx tsx ${adr}`;

  let adrDirectory: string;
  let workDir: string;

  beforeEach(async () => {
    // @ts-ignore
    process.env.ADR_DATE = '1992-01-12';
    workDir = path.resolve(realpathSync(await fs.mkdtemp(path.join(os.tmpdir(), 'adr-'))));
    adrDirectory = path.join(workDir, 'doc/adr');
  });

  afterEach(() => {
    vi.clearAllMocks();
    rmdirSync(workDir, {
      recursive: true,
      maxRetries: 3,
      retryDelay: 500
    });
  });

  it('should be able to supersede previous adrs', async () => {
    childProcess.execSync(`${command} new First Record`, { timeout: 10000, cwd: workDir });
    childProcess.execSync(`${command} new -s 1 Second Record`, { timeout: 10000, cwd: workDir });

    const first: string = path.join(adrDirectory, '0001-first-record.md');
    const second: string = path.join(adrDirectory, '0002-second-record.md');

    const firstContent = await fs.readFile(first, 'utf8');
    const secondContent = await fs.readFile(second, 'utf8');

    expect(firstContent).toMatchSnapshot();
    expect(secondContent).toMatchSnapshot();
  });

  it('should be able to supersede multiple records', async () => {
    childProcess.execSync(`${command} new First Record`, { timeout: 10000, cwd: workDir });
    childProcess.execSync(`${command} new Second Record`, { timeout: 10000, cwd: workDir });
    childProcess.execSync(`${command} new -s 1 -s 2 Third Record`, { timeout: 10000, cwd: workDir });

    const first: string = path.join(adrDirectory, '0001-first-record.md');
    const second: string = path.join(adrDirectory, '0002-second-record.md');
    const third: string = path.join(adrDirectory, '0003-third-record.md');

    const firstContent = await fs.readFile(first, 'utf8');
    const secondContent = await fs.readFile(second, 'utf8');
    const thirdContent = await fs.readFile(third, 'utf8');

    expect(firstContent).toMatchSnapshot();
    expect(secondContent).toMatchSnapshot();
    expect(thirdContent).toMatchSnapshot();
  });
});
