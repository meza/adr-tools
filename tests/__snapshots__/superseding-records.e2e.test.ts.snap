// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Superseding Adrs > should be able to supersede multiple records 1`] = `
"# 1. First Record

Date: 1992-01-12

## Status

Superseded by [3. Third Record](0003-third-record.md)

## Context

The issue motivating this decision, and any context that influences or constrains the decision.

## Decision

The change that we're proposing or have agreed to implement.

## Consequences

What becomes easier or more difficult to do and any risks introduced by the change that will need to be mitigated.
"
`;

exports[`Superseding Adrs > should be able to supersede multiple records 2`] = `
"# 2. Second Record

Date: 1992-01-12

## Status

Superseded by [3. Third Record](0003-third-record.md)

## Context

The issue motivating this decision, and any context that influences or constrains the decision.

## Decision

The change that we're proposing or have agreed to implement.

## Consequences

What becomes easier or more difficult to do and any risks introduced by the change that will need to be mitigated.
"
`;

exports[`Superseding Adrs > should be able to supersede multiple records 3`] = `
"# 3. Third Record

Date: 1992-01-12

## Status

Accepted

Supersedes [1. First Record](0001-first-record.md)

Supersedes [2. Second Record](0002-second-record.md)

## Context

The issue motivating this decision, and any context that influences or constrains the decision.

## Decision

The change that we're proposing or have agreed to implement.

## Consequences

What becomes easier or more difficult to do and any risks introduced by the change that will need to be mitigated.
"
`;

exports[`Superseding Adrs > should be able to supersede previous adrs 1`] = `
"# 1. First Record

Date: 1992-01-12

## Status

Superseded by [2. Second Record](0002-second-record.md)

## Context

The issue motivating this decision, and any context that influences or constrains the decision.

## Decision

The change that we're proposing or have agreed to implement.

## Consequences

What becomes easier or more difficult to do and any risks introduced by the change that will need to be mitigated.
"
`;

exports[`Superseding Adrs > should be able to supersede previous adrs 2`] = `
"# 2. Second Record

Date: 1992-01-12

## Status

Accepted

Supersedes [1. First Record](0001-first-record.md)

## Context

The issue motivating this decision, and any context that influences or constrains the decision.

## Decision

The change that we're proposing or have agreed to implement.

## Consequences

What becomes easier or more difficult to do and any risks introduced by the change that will need to be mitigated.
"
`;
